### PR TITLE
Lnd plan of attack

### PR DESCRIPTION
**Pull request by thomaseizinger**

### CODE REVIEW

File: 0022-lnd-plan-of-attack.adoc [hunk 1]
```diff
@@ -0,0 +1,318 @@
...
+:sectnums:
+:sectnumlevels: 5
+
+NOTE: Author: {authors} +
+Date: {revdate} +
+Tracking issue: https://github.com/comit-network/comit-rs/issues/1838[#1838]
+
+== Context
+
+We want to allow users to use Lightning for swapping assets.
```
Comment (bonomat) [inline 0022-lnd-plan-of-attack.adoc:14]: I think we can mention that this spike only covers LND. 

`HoldInvoice` is also supported by clightning but I we really can ignore this for now and focus only on LND. 

See BitcoinOptech Newsletter: https://bitcoinops.org/en/newsletters/2019/04/16/#c-lightning-2540

File: 0022-lnd-plan-of-attack.adoc [hunk 2]
```diff
@@ -0,0 +1,318 @@
...
+This allows us to specify the lightning protocol as a standalone protocol, without having to define, what happens on the other ledger.
+
+==== Messaging protocol
+
+We want to change our messaging protocol for various reasons.
+For further detail, please consult the communication protocol spike: https://github.com/comit-network/comit-rs/issues/1827
+In the context of this spike, the relevant points are:
+
+- The need for an ledger-specific protocol; the current messaging protocol only allows for a single protocol to be specified
+- Different information needs to be passed around for the HALight protocol; the current message format would make this more complicated than it needs to be as it would increase the polymorphism one level further (different bodies for different protocol combinations)
```
Comment (bonomat) [inline 0022-lnd-plan-of-attack.adoc:41]: HALight was not yet introduced. This confused me a bit until I read the rest of the document.

File: 0022-lnd-plan-of-attack.adoc [hunk 3]
```diff
@@ -0,0 +1,318 @@
...
+All the information is encoded in the HTTP body, which not only complicates the deserialization process but also makes it hard for clients to construct a correct payload.
+
+We can kill two birds with one stone if we choose to not add support for the HALight protocol on the existing endpoint but instead add new ones.
+
+. adding new routes is automatically backwards compatible and downstream projects can adapt them gradually
+. the new routes can take a different, less complicated design which aids error messages, deserialization code and client side implementations
+
+The proposal is roughly outlined here: https://github.com/comit-network/comit-rs/issues/1326.
+
+To start with, I would add the following routes: (bitcoin to LN seems pointless but of course we can still add it)
```
Comment (bonomat) [inline 0022-lnd-plan-of-attack.adoc:115]: That sounds a bit harsh :D Imho bitcoin to lightning will be a crucial feature in the future. There are already business offering these things as a service such as https://submarineswaps.org/ 

```suggestion
To start with, I would add the following routes: 
```

File: assets/0022-halight-eth-ln-swap.puml [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+par
+
+a->ac: Get swap
+ac-->a: Action: lnd-add-hold-invoice
+
+bc->eth: Watch for contract
+
+end
+
+a->al: Add hold invoice
```
Comment (bonomat) [inline assets/0022-halight-eth-ln-swap.puml:29]: What's the reasoning of having Alice to add the hold invoice and not Alice's CND? 

The downside of this is that some RPC commands such as add `addholdinvoice` are only available through the RPC interface and if you think back, it was really hard to get this working from the browser. 

Wouldn't it be easier if CND would just do this? 
I think letting CND doing this would cope with the race-condition issue you mention in `HALight-HAN swap` as the invoice can already added to LND by CND before the users received the invoice data.

File: assets/0022-halight-eth-ln-swap.puml [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+
+ac->al: Get invoice state
```
Comment (bonomat) [inline assets/0022-halight-eth-ln-swap.puml:31]: This looks like a request-response pattern but is actually a unidirectional stream:

https://github.com/lightningnetwork/lnd/blob/1dadce7cfd3ab20f4c50a9bb2600d3b11f146692/lnrpc/invoicesrpc/invoices.proto#L14-L18

Hence, I think we can just omit this line?

File: assets/0022-halight-ln-eth-swap.puml [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+
+a->eth: Call contract
+
+note right of a
+Invoking the contract reveals the secret!
+end note
+
+eth-->b: Contract was redeemed!
+
+b->b: Extract secret from transaction
```
Comment (bonomat) [inline assets/0022-halight-ln-eth-swap.puml]: Bob's CND should extract the secret right? Bob will then need to manually accept that payment via `settleInvoice`. 

```suggestion
bc->bc: Extract secret from transaction

b->bc: Get swap
bc-->b: Action: lnd-accept-invoice
```

File: assets/0022-halight-ln-eth-swap.puml [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+a->ac: Get swap
+ac-->a: Action: ethereum-call-contract
+
+a->eth: Call contract
+
+note right of a
+Invoking the contract reveals the secret!
+end note
+
+eth-->b: Contract was redeemed!
```
Comment (bonomat) [inline assets/0022-halight-ln-eth-swap.puml]: Imho Bob's CND should learn about this:

```suggestion
eth-->bc: Contract was redeemed!
```

File: 0022-lnd-plan-of-attack.adoc [hunk 8]
```diff
@@ -0,0 +1,318 @@
...
+
+The following diagrams assume that a negotiation protocol is in place and the parties agreed on a set of parameters.
+The protocol is following the "COMIT spirit":
+
+- readonly access to LND; COMIT is never in control of the user's funds
+- moving money is explicit; COMIT emits a dedicated `redeem` action for both protocol (even though it would not be necessary for the HAN-HALight one because Alice knows the secret from the very beginning so she could just use a regular invoice)
+
+===== HAN-HALight swap
+
+image::http://www.plantuml.com/plantuml/proxy?src=https://raw.githubusercontent.com/comit-network/spikes/700fd2c/assets/0022-halight-eth-ln-swap.puml&fmt=svg[HAN-HALight]
```
Comment (bonomat) [inline 0022-lnd-plan-of-attack.adoc]: Good job on those diagrams. I found a few minor things.

File: 0022-lnd-plan-of-attack.adoc [hunk 9]
```diff
@@ -0,0 +1,318 @@
...
+. Advantages:
+- Simple HTTP GET request
+
+. Disadvantages:
+- We have to poll regularly
+
+====== Using gRPC
+
+There is a dedicated gRPC call `SubscribeSingleInvoice` that allows to get a stream of changes of one invoice.
+This RPC call is behind the feature-flag `invoicesrpc` so unfortunately there is online documentation.
```
Comment (bonomat) [inline 0022-lnd-plan-of-attack.adoc]: 😅
```suggestion
This RPC call is behind the feature-flag `invoicesrpc` so unfortunately there is no online documentation.
```

File: 0022-lnd-plan-of-attack.adoc [hunk 10]
```diff
@@ -0,0 +1,318 @@
...
+
+===== `lnd-send-payment`
+
+===== `lnd-add-hold-invoice`
+
+===== `lnd-settle-invoice`
+
+== Recommendation
+
+- should we try and integrate lightning without changing our messaging??
```
Comment (bonomat) [inline 0022-lnd-plan-of-attack.adoc]: Imho yes. This would allow us to have a working feature soon on which we can then iterate. 
I know the downside of this is that _more_ code is harder to refactor.

File: assets/0022-halight-ln-eth-swap.puml [hunk 11]
```diff
@@ -0,0 +1,87 @@
...
+Alice cannot detect that she is paying a hold invoice.
+She doesn't know that Bob received the payment.
+end note
+
+ac->eth: Watch for contract
+
+b->bc: Get swap
+bc-->b: Action: ethereum-deploy-contract
+b->eth: Deploy Ethereum contract
+b->eth: Watch contract activity
```
Comment (bonomat) [inline assets/0022-halight-ln-eth-swap.puml]: Bob's CND should watch for the contract. 

```suggestion
bc->eth: Watch contract activity
```

File: 0022-lnd-plan-of-attack.adoc [hunk 12]
```diff
@@ -0,0 +1,318 @@
...
+
+- use HTTP for communication
+- start with adopting the two-ledger protocol style inside cnd with han and herc20
+- set up the test infrastructure with running lightning nodes and channel opened
+- start implementing the new communication protocol (other spike)
+- start with adding routes to cnd that can be called for these tests and write e2e tests that go as far as possible
```
Comment (bonomat) [inline 0022-lnd-plan-of-attack.adoc]: Is this an ordered list or overall recommendations?

**Review by bonomat — COMMENTED** [review]
Great write-up. I'm already super hyped about getting this implemented. 

You have it probably in your head already: can we have a few more details on in which order you see this being implemented please? 

I could think of something like this (if we follow your proposal of having most LND interactions outside of CND): 

1. Phase 0: e2e test lib support:
    * Create and redeem a hold invoice using JS in our e2e tests. We can start this in parallel of Phase 1: 
2. Phase 1: HALight-HAN - quick prototype: 
    * Implement LND client in CND.
    * Reuse messaging protocol 
    * Add new actions to cnd's Rest API (`lnd-send-payment`, `lnd-add-hold-invoice`, `lnd-settle-invoice`). 
    * Create e2e tests for HALight-HAN
3. Phase 2: Add new communication protocol ...
4. Phase 3:  
   * Extract things in libraries, etc. 
    * Add LND support in sdk (?we probably want this?)
    * Add example in create-comit-app (?we probably want this?)

File: 0022-lnd-plan-of-attack.adoc [hunk 13]
```diff
@@ -0,0 +1,448 @@
...
+
+=== Problems
+
+This section defines the problems we need to solve for a good integration of Lightning into our current software and specification.
+The problems identified are:
+
+. Swaps with Lightning are HTLC-based, yet they are different from our current swaps and we don't have a reusable concept in our spec and/or software for defining protocols per ledger
+. Different messaging is required for a swap that includes Lightning vs other protocols but our messaging is not flexible enough to express that
+. Lightning is not a public ledger, hence our current abstraction of watching a blockchain doesn't fit
+. The _standard_ LN as of today doesn't support paying someone who doesn't know the preimage, hence the swaps would only work in one direction (if Bob pays Alice via LN)
```
Comment (D4nte) [inline 0022-lnd-plan-of-attack.adoc]: Wait, isn't it what `hodlinvoice` is for?

File: 0022-lnd-plan-of-attack.adoc [hunk 14]
```diff
@@ -0,0 +1,448 @@
...
+. Our current REST API is not capable of the alpha/beta ledger protocol idea
```
Comment (D4nte) [inline 0022-lnd-plan-of-attack.adoc]:
```suggestion
. Our current REST API is not capable of using separate protocol for each ledger
```

File: 0022-lnd-plan-of-attack.adoc [hunk 15]
```diff
@@ -0,0 +1,448 @@
...
+==== Lightning is not a public ledger
+
+At the moment, the implementation of the rfc003 protocol is based on the `HtlcEvents` trait.
+Said trait exposes the events that can happen as part of the protocol execution.
+
+The implementations of that make use of the `btsieve` module which cannot be used for Lightning as already mentioned above.
+Hence, the integration of Lightning will likely require adding an "Events" traits, conceptually similar to `HtlcEvents`.
+`HtlcEvents` is already specific to a single ledger, so we can simply add a trait `halight::Events` that describes all the events that can happen during the execution of the HALight protocol.
+
+Note: `HtlcEvents` will be decomposed into single function traits with https://github.com/comit-network/comit-rs/issues/1779[#1779].
```
Comment (D4nte) [inline 0022-lnd-plan-of-attack.adoc:91]: I feel that with the introduction of HALight then the name `HtlcEvents` for HAN does not fit anymore.
Even with this new issue, I think the terminology `Htlc` is going to be confusing.
Maybe similarly to what you propose for HALight, we should update comit-network/comit-rs#1779 to create new traits which are `han::Funded` instead of `HtlcFunded`.

File: 0022-lnd-plan-of-attack.adoc (hunk 13 above)
Comment (D4nte) [reply 0022-lnd-plan-of-attack.adoc]: Ok now I see, maybe `standard LN` does not really apply.
I'd rephrase to 
```suggestion
. Most LN clients doesn't support paying someone who doesn't know the preimage by default, hence the swaps would only work in one direction (if Bob pays Alice via LN)
```
And even this rephrase does not capture the fact that lnd supports it with compilation flags.

File: 0022-lnd-plan-of-attack.adoc [hunk 16]
```diff
@@ -0,0 +1,448 @@
...
+
+We could opt to remove this redundancy and infer the ledger from the protocol.
+However, in my opinion, consistency wins here:
+
+. these URLs are unlikely to be typed by developers (as long as the use our SDKs)
+. in the future, I'd hope we expose siren actions for creating swaps, hence the developers only need to follow those actions and will therefore never see this URL anyway
+
+=== The HALight protocol
+
+The HALight protocol allows us to do atomic swaps using HTLCs with assets that are tracked on lightning.
```
Comment (D4nte) [inline 0022-lnd-plan-of-attack.adoc]:
```suggestion
The HALight protocol allows us to do atomic swaps using HTLCs with assets that are tracked on Lightning.
```

File: 0022-lnd-plan-of-attack.adoc [hunk 17]
```diff
@@ -0,0 +1,448 @@
...
+One could conceptualize a single Lightning channel as a ledger in the way that it tracks ownership of who owns how many Bitcoins.
+Using a channel would be more correct than using a single node but bears other problems.
+In particular, payments in Lightning seldomly involve a single channel.
+Using the channel = Ledger model would work if there were not hops involved within a payment in Lightning.
+
+===== Why referring to the whole Lightning network as the Ledger
+
+It is conceptually similar to how we see other ledgers (Bitcoin, Ethereum).
+The difference is that it is a private ledger, hence not every party can see every transaction.
+I feel the concept of a "private" ledger is better suited here.
```
Comment (D4nte) [inline 0022-lnd-plan-of-attack.adoc:169]: You can add that this creates some synergy with Grin?

File: 0022-lnd-plan-of-attack.adoc [hunk 18]
```diff
@@ -0,0 +1,448 @@
...
+
+The following diagrams assume that a negotiation protocol is in place and the parties agreed on a set of parameters.
+The protocol is following the "COMIT spirit":
+
+- readonly access to LND; COMIT is never in control of the user's funds
+- moving money is explicit; COMIT emits a dedicated `redeem` action for both protocol (even though it would not be necessary for the HAN-HALight one because Alice knows the secret from the very beginning so she could just use a regular invoice)
+
+===== HAN-HALight swap
+
+image::http://www.plantuml.com/plantuml/proxy?src=https://raw.githubusercontent.com/comit-network/spikes/3e474ff2/assets/0022-halight-eth-ln-swap.puml&fmt=svg[HAN-HALight]
```
Comment (D4nte) [inline 0022-lnd-plan-of-attack.adoc:187]: s/protcol/protocol/

Comment (D4nte) [reply 0022-lnd-plan-of-attack.adoc:187]: Interesting that you decided to use hodl invoice for both directions. Why?

File: 0022-lnd-plan-of-attack.adoc [hunk 19]
```diff
@@ -0,0 +1,448 @@
...
+
+===== Watching an invoice
+
+We need to be able to differentiate between the following states:
+
+* Hold invoice added to LND but not paid yet (see <<HAN-HALight swap>>, cnd needs to be able detect that the invoice was added to emit the next action)
+* Invoice has been paid (analogous to HTLC has been deployed)
+* Invoice has been settled (analogous to HTLC has been redeemed)
+* Invoice has been rejected (this does not exist on Layer 1, it would be analogous for removing the HTLC from the blockchain again)
+* Invoice has timed out
```
Comment (D4nte) [inline 0022-lnd-plan-of-attack.adoc]: Can this happen if there is a pending payment on such invoice? How much do we control that?

File: 0022-lnd-plan-of-attack.adoc [hunk 20]
```diff
@@ -0,0 +1,448 @@
...
+. secret-hash
+. expiry:
+This allows us to specify, for how long we are willing to accept a payment for this invoice.
+The default here is 3600 seconds, e.g. 1 hour.
+We should return sensible values here, depending on the roles and the protocol that is used on the other chain.
+If this timeout is too short, Bob might not be able to pay the invoice in the X-HALight case even though we would like to proceed with the swap.
+For the HALight-X case, this would allow Bob to specify a time window up until he would like Alice to start the swap.
+In this case, the timeout should not be too long, otherwise the absolute expiry of his HTLC (assuming a HAN-protocol) is not enough for a secure swap.
+. memo
+We can return a description that says something like "COMIT atomic swap {swap-id} with {peer-id}" to make identifying invoices of swaps easier for the user.
```
Comment (D4nte) [inline 0022-lnd-plan-of-attack.adoc:351]: FYI, max size is 639 bytes (I guess we are fine) https://github.com/lightningnetwork/lightning-rfc/blob/master/11-payment-encoding.md#tagged-fields

File: 0022-lnd-plan-of-attack.adoc [hunk 21]
```diff
@@ -0,0 +1,448 @@
...
+. The support for Lightning inside cnd
+. The messaging protocol
+
+=== The e2e test suite
+
+To know when we are done with integrating Lightning, we should start with extending our e2e test suite with the respective tests.
+Our e2e test suite is in a limbo state at the moment, hence the first step is to get this sorted.
+
+. Make the e2e tests run in parallel: https://github.com/comit-network/comit-rs/issues/1927 (including the dependencies)
+. Setup up the test infrastructure to start an LND instance per actor and open a channel between them
```
Comment (D4nte) [inline 0022-lnd-plan-of-attack.adoc:383]: It would be interesting to add an edge case with `Alice -- Charlie -- Bob` channel topography.

File: 0022-lnd-plan-of-attack.adoc [hunk 22]
```diff
@@ -0,0 +1,448 @@
...
+=== The messaging protocol
+
+This spike doesn't go into detail on what the new communication protocol looks like.
+However, I'd recommend to not follow an iterative approach but rather implement it as a separate crate on the side and drop it in as a replacement for the current protocol once it is ready.
+Once the new communication protocol is in, we can add the necessary messages for Lightning to make more progress on that.
+
+Given the above recommendations, we can specify the following requirements for the new communication protocol:
+
+. Both parties need to know about the pubkeys of their lnd instances
+. Both parties need knowledge about the final cltv delta
```
Comment (D4nte) [inline 0022-lnd-plan-of-attack.adoc:415]: I believe this is configurable on lnd. It'd be good to clarify this part.

**Review by D4nte — COMMENTED** [review]
Overall comment:
- What changes on the SDK? e.g. `Adding a hold invoice is for now only possible via the gRPC API.` which means we need to bring gRPC to the SDK?

File: 0022-lnd-plan-of-attack.adoc (hunk 22 above)
Comment (D4nte) [reply 0022-lnd-plan-of-attack.adoc:415]: This should be verified as part of the `haLight::events` (deployed).

File: 0022-lnd-plan-of-attack.adoc (hunk 15 above)
Comment (thomaseizinger) [reply 0022-lnd-plan-of-attack.adoc:91]: It really depends on which issue happens first.
I'd expect everyone to follow along here good enough to get this :)

File: 0022-lnd-plan-of-attack.adoc (hunk 2 above)
Comment (thomaseizinger) [reply 0022-lnd-plan-of-attack.adoc:41]: I assumed readers of this spike are aware of all previously written spikes.

File: 0022-lnd-plan-of-attack.adoc (hunk 3 above)
Comment (thomaseizinger) [reply 0022-lnd-plan-of-attack.adoc:115]: Just because there are business offerings doesn't mean it is not pointless :)

Just trying to reduce the scope here but we can ofc add any combination.

Comment (thomaseizinger) [reply 0022-lnd-plan-of-attack.adoc:115]: BTC to LN is also not cross-chain. We have this discussion at some point and I remember we agreed that it is not a priority for us?

File: assets/0022-halight-eth-ln-swap.puml (hunk 4 above)
Comment (thomaseizinger) [reply assets/0022-halight-eth-ln-swap.puml:29]: > What's the reasoning of having Alice to add the hold invoice and not Alice's CND?

Simplicity for now. See last section I added.

> The downside of this is that some RPC commands such as add addholdinvoice are only available through the RPC interface and if you think back, it was really hard to get this working from the browser.

If we want it from the browser I suggest we wait until it is stabilized and an HTTP interface is provided.
I am not aware of any browser-based COMIT projects in the near future.
Given that this problem will solve itself over time, I opted for the easiest solution for now.

File: assets/0022-halight-eth-ln-swap.puml (hunk 5 above)
Comment (thomaseizinger) [reply assets/0022-halight-eth-ln-swap.puml:31]: The recommendation says use HTTP so it is actually request-response :)

File: 0022-lnd-plan-of-attack.adoc (hunk 10 above)
Comment (thomaseizinger) [reply 0022-lnd-plan-of-attack.adoc]: I thought about this quite a bit but decided against it. See the last section for details.

**Review by luckysori — COMMENTED** [review]
Excellent work. Nothing to report, it looks good to me.

File: 0022-lnd-plan-of-attack.adoc (hunk 3 above)
Comment (bonomat) [reply 0022-lnd-plan-of-attack.adoc:115]: No worries, just thought it is not pointless :)

I'm thinking about various use cases for some time now and how to get COMIT known in a community we feel comfortable with. That's why I'd consider BTC<->LN and USDT<->LN. 
... Ithought I add some more insights in my thinking. Please feel free to leave it as is.

File: 0022-lnd-plan-of-attack.adoc (hunk 21 above)
Comment (bonomat) [reply 0022-lnd-plan-of-attack.adoc:383]: Good point. Maybe don't even call it edge case but default. At the end I think it is more likely that there is someone between Alice and Bob than that there is a direct connection between the two.

**Review by D4nte — APPROVED** [review]

**Review by luckysori — APPROVED** [review]

**Review by tcharding — APPROVED** [review]

**Comment by tcharding** [comment]
Pushed two minor patches, one grammar, one rendering.

File: 0022-lnd-plan-of-attack.adoc (hunk 2 above)
Comment (bonomat) [reply 0022-lnd-plan-of-attack.adoc:41]: That's fine. For the future I would either link to the other spike or shortly introduce the abbreviation here as well.

File: assets/0022-halight-eth-ln-swap.puml (hunk 4 above)
Comment (bonomat) [reply assets/0022-halight-eth-ln-swap.puml:29]: That's fine. We should keep this justification in mind next time when we start discussing: "why isn't cnd doing this for me".

File: assets/0022-halight-eth-ln-swap.puml (hunk 5 above)
Comment (bonomat) [reply assets/0022-halight-eth-ln-swap.puml:31]: Got it.

File: assets/0022-halight-ln-eth-swap.puml [hunk 23]
```diff
@@ -0,0 +1,90 @@
...
+a->al: Pay invoice
+
+note right a
+You might notice a potential race-condition here. (We are in a par block!)
+What if Alice tries to pay the invoice before Bob added it do his LND node?
+
+The answer to this is "try again". We already make an assumption that cnd's
+clients can successfully make transactions.
+Failing to make a payment because the invoice does not yet exist is similar
+to failing to make a Bitcoin transaction because the fee was to low.
```
Comment (bonomat) [inline assets/0022-halight-ln-eth-swap.puml:37]: If the fee was too low the transaction would remain pending for a long time (some nodes never clear the mempool).

```suggestion
to failing to make a Bitcoin transaction because it was invalid.
```

**Review by bonomat — APPROVED** [review]
Thanks for writing this up. I added a few minor comments but overall good work. 

For the next spike/document, can you use 1-sentence-per-line approach again please? I find this easier to read/review. 

Note: I ignored grammar/typos completely :)

**Comment by tcharding** [comment]
I merged this @thomaseizinger, please yell if this messes with your flow.

**Comment by thomaseizinger** [comment]
@bonomat, I've actually tried to follow the one sentence per line thingy. Apologies if I've missed it somewhere :)
